src/main/java/io/r2dbc/postgresql/UniformLoadBalancerConnectionStrategy.java

```diff
@@ -0,0 +1,166 @@
+    }
+
+
+    protected List<String> getPrivateOrPublicServers(List<String> privateHosts,
+                                                          List<String> publicHosts) {
+        if (useHostColumn == null) {
+            if (publicHosts.isEmpty()) {
+                useHostColumn = Boolean.TRUE;
+            }
+            return privateHosts;
+        }
+        List<String> currentHosts = useHostColumn ? privateHosts : publicHosts;
+        return currentHosts;
+    }
+
+    @Override
+    public Mono<Client> connect() {
+        return null;
+    }
+
+    public Mono<Client> connect(String host) {
+        incDecConnectionCount(host, 1);
+        endpoint = InetSocketAddress.createUnresolved(host, 5433);
+        Mono<Client> client = this.connectionFunction.connect(endpoint, this.connectionSettings);
+        return client;
+    }
+
+    public synchronized void updateFailedHosts(String chosenHost) {
+        unreachableHosts.add(chosenHost);
+        hostToConnectionCount.remove(chosenHost);
+    }
+
+    public boolean needsRefresh() {
+        long currentTimeInMillis = System.currentTimeMillis();
+        long diff = (currentTimeInMillis - lastServerListFetchTime) / 1000;
+        boolean firstTime = servers == null;
+        if (firstTime || diff > refreshListSeconds) {
+            return true;
+        }
+        return false;
+    }
+
+    public synchronized boolean refresh(PostgresqlConnection controlConnection) {
+        if (!needsRefresh()) {
+            return true;
+        }
+        // else clear server list
+        long currTime = System.currentTimeMillis();
+        lastServerListFetchTime = currTime;
+        long now = System.currentTimeMillis() / 1000;
+
+        servers = getCurrentServers(controlConnection);
+        unreachableHosts.clear();
+        if (servers == null) {
+            return false;
+        }
+
+        for (String h : servers) {
+            if (!hostToConnectionCount.containsKey(h)) {
+                hostToConnectionCount.put(h, 0);
+            }
+        }
+        return true;
+    }
+
+    public void setForRefresh() {
+        lastServerListFetchTime = 0L;
+    }
+
+    // When a connection is closed, decrement the connection count for the host
+
+    public void incDecConnectionCount(String host, int incDec) {
+        if (hostToConnectionCount.get(host) == 0 && incDec < 0)
+            return;
+        hostToConnectionCount.put(host, hostToConnectionCount.get(host) + incDec);
+    }
+
+    // Get the host with the least number of connections
+    public String getHostWithLeastConnections() {
+        String hostWithLeastConnections = "";
+        int leastConnections = Integer.MAX_VALUE;
+        for (String host : hostToConnectionCount.keySet()) {
+            if (!unreachableHosts.contains(host) && hostToConnectionCount.get(host) < leastConnections) {
+                hostWithLeastConnections = host;
+                leastConnections = hostToConnectionCount.get(host);
```

Comment:
What if there are Multiple hosts with Least Connections?

Reply:
They get randomly picked in that case.